src/devices/Gpio/Drivers/Sunxi/SunxiDriver.cs

```diff
@@ -0,0 +1,387 @@
+                        SetPinMode(pinNumber, PinMode.Input);
+                        break;
+                    default:
+                        break;
+                }
+
+                _pinModes.Remove(pinNumber);
+            }
+        }
+
+        /// <inheritdoc/>
+        protected override void SetPinMode(int pinNumber, PinMode mode)
+        {
+            // Get port controller, port number and shift
+            (int PortController, int Port) unmapped = UnmapPinNumber(pinNumber);
+            int cfgNum = unmapped.Port / 8;
+            int cfgShift = unmapped.Port % 8;
+            int pulNum = unmapped.Port / 16;
+            int pulShift = unmapped.Port % 16;
+
+            // Pn_CFG is used to set the direction; Pn_PUL is used to set the pull up/dowm mode
+            uint* cfgPointer, pulPointer;
+            int cfgOffset, pulOffset;
+
+            if (unmapped.PortController < 10)
+            {
+                // Pn_CFG initial offset is 0x00
+                cfgOffset = (CpuxPortBaseAddress + unmapped.PortController * 0x24 + cfgNum * 0x04) & _mapMask;
+                // Pn_PUL initial offset is 0x1C
+                pulOffset = (CpuxPortBaseAddress + unmapped.PortController * 0x24 + 0x1C + pulNum * 0x04) & _mapMask;
+
+                cfgPointer = (uint*)(_cpuxPointer + cfgOffset);
+                pulPointer = (uint*)(_cpuxPointer + pulOffset);
+            }
+            else
+            {
+                cfgOffset = (CpusPortBaseAddress + unmapped.PortController * 0x24 + cfgNum * 0x04) & _mapMask;
+                pulOffset = (CpusPortBaseAddress + unmapped.PortController * 0x24 + 0x1C + pulNum * 0x04) & _mapMask;
+
+                cfgPointer = (uint*)(_cpusPointer + cfgOffset);
+                pulPointer = (uint*)(_cpusPointer + pulOffset);
+            }
+
+            uint cfgValue = *cfgPointer;
+            uint pulValue = *pulPointer;
+
+            // Clear register
+            // Input is 0b000; Output is 0b001
+            cfgValue &= ~(0b1111U << (cfgShift * 4));
+            // Pull-up is 0b01; Pull-down is 0b10; Default is 0b00
+            pulValue &= ~(0b11U << (pulShift * 2));
+
+            switch (mode)
+            {
+                case PinMode.Output:
+                    cfgValue |= 0b_001U << (cfgShift * 4);
+                    break;
+                case PinMode.Input:
+                    // After clearing the register, the value is the input mode.
+                    break;
+                case PinMode.InputPullDown:
+                    pulValue |= 0b10U << (pulShift * 2);
+                    break;
+                case PinMode.InputPullUp:
+                    pulValue |= 0b01U << (pulShift * 2);
+                    break;
+                default:
+                    throw new ArgumentException("Unsupported pin mode.");
+            }
+
+            *cfgPointer = cfgValue;
+            *pulPointer = pulValue;
+
+            if (_pinModes.ContainsKey(pinNumber))
```

Comment:
consider using TryGetValue

Reply:
To eliminate this warning, I need extra steps to deal with the situation where pinState is null (although it cannot be null). 😅
```C#
//  VS warns me that pinState may be null
if (_pinModes.TryGetValue(pinNumber, out PinState pinState))
{
}
```

Reply:
I'm not sure what you mean but fine as is